Add WENO limiter specialized to the NewtonianEuler system

fmahebert opened the pull request:
## Proposed changes

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Add WENO limiter specialized to NewtonianEuler

fmahebert committed:
Fix Doxygen documentation for Weno limiter

By moving the Doxygen from the template specialization to the
declaration, we help Doxygen understand how to link to this
documentation from elsewhere in the code.

fmahebert committed:
Update Kxrcf TCI for new normal vectors

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+                 typename ConservativeVarsWeno::DisableForDebugging>;
+  static constexpr Options::String help = {
+      "A WENO limiter specialized to the NewtonianEuler system"};
+  static std::string name() noexcept { return "NewtonianEulerMinmod"; };
```
This name needs updating, I think

fmahebert replied:
I agree 😛

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+  // is also used when limiting cons vars with the simple WENO algorithm. So we
+  // construct conservative_vars_weno_ with the correct TVB constant when it is
+  // used for limiting (precisely, when weno_type_ == SimpleWeno), and with a
+  // dummy TVB constant value of 0.0 otherwise (weno_type_ == Hweno). This lets
```
would setting to `signaling_NaN()` instead of `0.0` work?

fmahebert replied:
It does 👍

nilsdeppe commented on tests/Unit/Evolution/Systems/NewtonianEuler/Limiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,805 @@
+  // make sense... otherwise we aren't testing anything useful
+  const bool has_negative_density = min(get(input_density)) < 0.;
+  REQUIRE(has_negative_density);
+  CHECK(true);
```
remove?

nilsdeppe reviewed (changes requested):
You can squash immediately

fmahebert commented:
Squashed and pushed the fixes from @nilsdeppe's review.

In addition to what was requested, I fixed a bug in the parsing of the optional inputs in the weno constructor (the cpp file line 332ish and below)

nilsdeppe reviewed (dismissed):
LGTM (including the constructor changes)

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+///
+/// Implements the simple WENO limiter of \cite Zhong2013 and the Hermite WENO
+/// (HWENO) limiter of \cite Zhu2016. See the documentation of the system-
+/// agnostic ::Limiters::Weno limiter for a general discussion of the algorithm
```
there is a space showing up between `system-` and `agnostic`

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+                 typename ConservativeVarsWeno::DisableForDebugging>;
+  static constexpr Options::String help = {
+      "A WENO limiter specialized to the NewtonianEuler system"};
+  static std::string name() noexcept { return "NewtonianEulerWeno"; };
```
`#include <string>`

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+
+  Weno(::Limiters::WenoType weno_type,
+       NewtonianEuler::Limiters::VariablesToLimit vars_to_limit,
+       double neighbor_linear_weight, std::optional<double> tvb_constant,
```
`#include <optional>`

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+      const tnsr::I<DataVector, VolumeDim>& momentum_density,
+      const Scalar<DataVector>& energy_density, const Mesh<VolumeDim>& mesh,
+      const std::array<double, VolumeDim>& element_size,
+      const OrientationMap<VolumeDim>& orientation_map) const noexcept;
```
forward include for OrientationMap

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+#include <unordered_map>
+#include <utility>
+
+#include "DataStructures/Tensor/Tensor.hpp"
```
just need TypeAliases.hpp

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+  /// \brief Package data for sending to neighbor elements
+  void package_data(
+      gsl::not_null<PackagedData*> packaged_data,
+      const Scalar<DataVector>& mass_density_cons,
```
forward declare DataVector

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+#include "Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp"
+#include "Evolution/Systems/NewtonianEuler/Limiters/VariablesToLimit.hpp"
+#include "Evolution/Systems/NewtonianEuler/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
```
just need to forward declare

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Options/Auto.hpp"
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
```
just need to forward declare

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "PointwiseFunctions/Hydro/TagsDeclarations.hpp"
+#include "Utilities/Gsl.hpp"
```
just need to forward declare

kidder replied:
when squashing you can remove this include since you added a forward declaration

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.hpp:
```diff
@@ -0,0 +1,205 @@
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/SizeOfElement.hpp"
+#include "Domain/Structure/Element.hpp"  // IWYU pragma: keep
```
just need to forward declare

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+namespace {
+template <size_t VolumeDim, size_t ThermodynamicDim>
+bool characteristic_simple_weno_impl(
+    const gsl::not_null<Scalar<DataVector>*> mass_density_cons,
```
`#include "DataStructres/DataVector.hpp"`

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+    const double tvb_constant, const double neighbor_linear_weight,
+    const Mesh<VolumeDim>& mesh, const Element<VolumeDim>& element,
+    const std::array<double, VolumeDim>& element_size,
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
```
#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
+        equation_of_state,
+    const std::unordered_map<
+        std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>,
```
includes for Direction and ElmentId

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+  // Storage for transforming neighbor_data into char variables
+  using CharacteristicVarsWeno =
+      Limiters::Weno<VolumeDim,
+                     tmpl::list<NewtonianEuler::Tags::VMinus,
```
`#include Utilities/TMPL.hpp"`

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+  // Buffers for SimpleWeno extrapolated poly
+  std::unordered_map<
+      std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>,
+      intrp::RegularGrid<VolumeDim>,
```
include for RegularGrid

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+        std::make_pair(neighbor, DataVector(mesh.number_of_grid_points())));
+  }
+
+  // Outer lambda: wraps applying Hweno to the NewtonianEuler
```
there is no inner lambda, so drop `Outer lambda: `

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Weno.cpp:
```diff
@@ -0,0 +1,553 @@
+    const ::Limiters::WenoType weno_type,
+    const NewtonianEuler::Limiters::VariablesToLimit vars_to_limit,
+    const double neighbor_linear_weight,
+    const std::optional<double> tvb_constant,
```
`#include <optional>`

kidder commented on tests/Unit/Evolution/Systems/NewtonianEuler/Limiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,804 @@
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
```
check for any missing includes (likely the same I flagged for the source files)

fmahebert commented:
Pushed a fixup that cleans up the includes

kidder reviewed (commented):
you can squash (making the one extra change)

fmahebert commented:
Squashed the fixup (including deleting the include pointed out by @kidder) and rebased.
